### ISSUE TITLE
Why is glacier using only 160 bits of entropy per key?

### ISSUE THREAD

**Issue by jazarija** [GlacierProtocol/GlacierProtocol#48]
As I currently understand the protocol, Glacier asks the user to generate 160 random bits (via dice rolls) for each private key.  This is then mixed with an additional 160 bits taken from /dev/random.

Now, a BTC private key has a size of 256 bits, and the process above seems to ignore almost 96 bits, drastically reducing the search space for private keys used by Glacier.

Why is this a good idea? 

Unless I am missing something, glacier is deliberately generating private keys that are almost 50% weaker than what they are supposed to be?

**Comment by bitcoinhodler** [comment]
There's a note there in the Glacier doc when explaining the dice rolls: "62 dice rolls corresponds to 160 bits of entropy. See the design document for details."

In the design document we find:

> Bitcoin addresses are 160-bit hashes of a 256-bit public key. Any keypair which has a public key that
hashes to an address can spend the coins at that address. Therefore, the most entropy that is useful is
160 bits. This corresponds to 62 dice rolls.

Though in this case it's P2SH, so not exactly like this, but still only 160 bits.

Also remember you have N pubkeys in the script, so if you're doing 2-of-4 you have 4*160 bits of entropy in the redeem script.

If we upgraded to P2WSH (bech32) then the address would be 256 bits, but not a lot of wallets are compatible with that yet, and in any case, 160 bits is plenty for the foreseeable future.

**Comment by jazarija** [comment]
Thank you, I am learning new stuff everyday!

As a follow up question, would you say the script should check if the produced key is smaller than FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F as described [here](https://en.bitcoin.it/wiki/Secp256k1)?

**Comment by bitcoinhodler** [comment]
I've considered that. The valid range is explained in the [Bitcoin wiki](https://en.bitcoin.it/wiki/Private_key) as follows:

> Nearly every 256-bit number is a valid ECDSA private key. Specifically, any 256-bit number from 0x1 to 0xFFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFE BAAE DCE6 AF48 A03B BFD2 5E8C D036 4140 is a valid private key. 

It's true that GlacierScript will fail if the randomly generated key is outside the legal ECDSA range. But since the legal range is such a large percentage of the entire 2^256 range, the odds of ever generating a key outside the range are astronomically low. The first 127 bits would have to be all 1's, which means less than 1 in 2^127.

Making it worse, how would we test such code? We'd have to find a combination of entropy that resulted in an illegal key value, which is practically impossible.

Bottom line: it will never happen, so it's not something we need to plan for.
